### PR TITLE
Add parameterized query wrapper

### PR DESCRIPTION
**Pull request by diacus**
## What does this PR do?

As part of the SQL injection security feature, this patch introduces a SQL query wrapper to control how the query parameters are interpolated in the sentence.

## Background context

See [this ticket](https://wizeline.atlassian.net/browse/BAD-12)

## Where should the reviewer start?

- `bigquery/query_builder.py`

## How can it be manually tested?

1. Use a new virtual environment
1. Install the package requirements
1. Run `python -m unittest`

## What are the relevant tickets

- See [BAD-12](https://wizeline.atlassian.net/browse/BAD-12)

## Does this change require to update the documentation on README.md?
No

### CODE REVIEW

File: bigquery/query_builder.py [hunk 1]
```diff
@@ -1,9 +1,36 @@
...
+    def _escape_single_quotes(self, arguments):
+        return {
+            arg_name: self._escape_string_value(arg_value)
+            for arg_name, arg_value in arguments.items()
+        }
+
+    def _escape_string_value(self, value):
+        try:
+            return re.sub(UNESCAPED_QUOTE, _replace_quote_with, value)
+        except TypeError:
```
Comment (juanitodread) [inline bigquery/query_builder.py:40]: Are you sure that only `TypeError` can be raised here? If another error/exception is raised, it may break the execution.

Would it be better to verify the type instead? Something like:

```python
def _escape_string_value(self, value):
    if type(value) == str:
        return re.sub(UNESCAPED_QUOTE, _replace_quote_with, value)
    else:
        return value
```

File: bigquery/query_builder.py [hunk 2]
```diff
@@ -397,3 +424,7 @@ def _render_limit(limit):
         return ''
 
     return "LIMIT %s" % limit
+
+
+def _replace_quote_with(match):
+    return match.group(0)[0] + r"\'"
```
Comment (juanitodread) [inline bigquery/query_builder.py]: 👀 this is a little bit hard to get it

**Review by juanitodread — COMMENTED** [review]
Does this functionality already covers _SQLInjection_ attacks?

File: bigquery/query_builder.py (hunk 1 above)
Comment (diacus) [reply bigquery/query_builder.py:40]: Yes, the Type error is raised when `value` is not a string,  a byte sequence, nor Pattern object. That's why I chose the `try/except` approach here. I think the code gets shorter. The reasoning is this:

1. First try to substitute the pattern
1. If the pattern is not found, then we return the same value
1. if the patter is found, then we return the _escaped_ value
1. if the substitution fails (it always will raise TypeError, since we're not doing weird stuff like passing flags to `re.sub()`

File: bigquery/query_builder.py (hunk 2 above)
Comment (diacus) [reply bigquery/query_builder.py]: That's why it is wrapped in function with a descriptive name: the function replaces a match with te string returned by it

**Comment by diacus** [comment]
By now it just prevents to break the queries by unescaped quote characters, in following iterations it will prevent SQL injection

File: bigquery/query_builder.py (hunk 1 above)
Comment (juanitodread) [reply bigquery/query_builder.py:40]: Ok, it's fine. Since in the tests you are testing just the integer case... Do you think makes sense to verify other scenarios (different types)? So, that our tests are going to guarantee those types are handled in the `except`.

File: bigquery/query_builder.py (hunk 2 above)
Comment (juanitodread) [reply bigquery/query_builder.py]: Yes, the name of the function is well descriptive. I said that because of the param and the body... 

Is `match` a collection or a string or a regex or what is? 
What does `group(0)[0]` do here? 
Why `(0)[0]`? 
What does `r"\'"` means?. 

You know, the code is not so explicit (not because you, but because the way these regex functions are implemented in the language). If a new developer has to modify this function, at least he/she should has a good understanding of regex API, to make sure everything continue working as desired after modification.

Comment (diacus) [reply bigquery/query_builder.py]: Got it, I will add a comment.

File: bigquery/query_builder.py (hunk 1 above)
Comment (diacus) [reply bigquery/query_builder.py:40]: I can add test agains `None`, `object()` and other collections, just to be sure :)

File: bigquery/query_builder.py (hunk 2 above)
Comment (n-sanchez) [reply bigquery/query_builder.py]: Perhaps you can also make this regex a constant, like you do above with UNESCAPED_QUOTE, just to make things a little easier to read.

**Review by n-sanchez — APPROVED** [review]
I think everything looks pretty good 👍
